### DISCUSSION TITLE
How to delete a row by Execute Query

### DISCUSSION THREAD

**Discussion by dungtrantt** [Budibase/budibase#15202] [Help]
Hi all,
I wanted to implement as below steps:
* Load data from REST API to Super table
* Each row in Super table, I have Delete and Edit button
* Edit button: When clicking on it, it shows an Modal up with info of the clicked row
* Delete button: When clicking on it, it should execute an API to delete that row by id of that record

My problems are:
* For Edit button, I binded from CurrentRow in Super table to form fields in Modal, during designing, data shows correctly, but in Preview, it didn't work.
* For Delete button, I did similar to Edit that I binded id field of CurrentRow in Super table to my API request, but it didn't work also

I seems data of CurrentRow couldn't show up in Modal or send to Action of Delete button.
Does anyone have any idea for this case?

**Comment by mikesealey** [comment]
Hey @dungtrantt 

The Super Table is a custom plugin created by a pillar of the Budibase community, @poirazis. He may be able to offer you more specific help.

For the time being, I can ask a few questions;
Do you have an endpoint on the REST API connector that will allow you to edit? 
Do you have an endpoint on the REST API connector that will allow you to delete?
Do you then have a Refresh Data Provider action to be triggered _after_ you make the changes?

For this last point, you can narrow down the issue slightly by making the change (edit/delete); If that change is not reflected in the app until you refresh the page, you can remedy it by adding a Refresh Data Provider.

You should also check what you're sending to your API requests by using the values in a Show Notification step (just while debugging, be sure to delete it before you send to procuction). You may find that you're accidentally sending the entire row, rather than a row_id.

**Reply by dungtrantt** [reply]
@mikesealey Thank you for your info. Below is information for your questions:

Do you have an endpoint on the REST API connector that will allow you to edit? => Yes. I did test successfully from Budibase in Data connector tab
Do you have an endpoint on the REST API connector that will allow you to delete? Yes. I did test successfully from Budibase in Data connector tab
Do you then have a Refresh Data Provider action to be triggered after you make the changes? Yes. I added this for Edit and Delete button also. Actually, I have another button out of Super table to create new item, I have a Refresh Data Provider  action after executing query of API creation. It works well.

I just tried with a Show Notification, but it seems it showed nothing.
Meantime, I submitted an issue to @poirazis as well. Hope we can hear from him soon.

**Comment by poirazis** [comment]
Hello @dungtrantt 
i think the problem here is the ST cannot identify which column from the REST query to use as an id column.
Does the REST query contain an "_id" or "id" field ? 
If no, can you try setting one ?

@mikesealey Thank you for your kind words !

**Reply by dungtrantt** [reply]
@poirazis Hi. We have id field. I used that, but it didn't work

**Reply by poirazis** [reply]
can you try the latest pushed version ? @dungtrantt

**Reply by poirazis** [reply]
btw , the correct binding name is` {{ New Super Table.Current Row }}`

**Reply by dungtrantt** [reply]
@poirazis I just tried with latest version. I recreated a Super Table component in the page after adding that version. During binding, data showed, but when I tested in Preview, it didn't work
<img width="1139" alt="Screenshot 2024-12-19 at 12 47 08" src="https://github.com/user-attachments/assets/1b65f601-08c6-47da-b8da-8a6e8b02ad03" />
<img width="786" alt="Screenshot 2024-12-19 at 12 48 20" src="https://github.com/user-attachments/assets/93c728d2-e710-4350-9944-a89bfa32572d" />
<img width="801" alt="Screenshot 2024-12-19 at 12 43 48" src="https://github.com/user-attachments/assets/267a3400-f03b-4f5a-ba1d-4f9b9184525c" />

**Reply by dungtrantt** [reply]
@poirazis One more update, I did try to add a Show Notification before Execute Query to print {{ New Super Table.Current Row }}. i saw whole record even id field, but in next action with Execute Query, I binded {{ New Super Table.Current Row.id }}, it couldn't work. Note: I was trying for Delete button

**Comment by poirazis** [comment]
hey @dungtrantt , i am having difficulty reproducing it.
can you give me an app export ( with another REST one of the open ones on the internet ) ?

**Comment by dungtrantt** [comment]
@poirazis We don't have REST APIs which are ready on the internet right now, but I will try to reproduce something similar. 
BTW, I checked Show Notification to print only id, it showed up. It seems it just can not put it into Execute Query, I checked Inspect, the request was sent with any id = "".

**Comment by dungtrantt** [comment]
@poirazis Below is an app which I tried to copy the same case. I setup an Open button at each row, this would execute REST API to each post. The query result should be show in a Show Notification after clicking Open button. But, yes, it still doesn't work as expected.
[Test App Super Table-export-1734853334970.enc.tar.gz](https://github.com/user-attachments/files/18221117/Test.App.Super.Table-export-1734853334970.enc.tar.gz)

**Reply by poirazis** [reply]
@dungtrantt a temporary work around until i solve is would be to use a Set State action , put the Current Row id there, and then bind the Execute Query with the State variable..

**Reply by dungtrantt** [reply]
@poirazis I just tried with Set State action. But, it still has the same issue. 
<img width="785" alt="Screenshot 2024-12-22 at 13 44 53" src="https://github.com/user-attachments/assets/1ab3944c-26c9-4060-b22f-ec557226b472" />
<img width="794" alt="Screenshot 2024-12-22 at 13 45 07" src="https://github.com/user-attachments/assets/7502b9ac-b44f-40b4-a489-1991f5633dcd" />

[Test App Super Table-export-1734893185470.enc.tar.gz](https://github.com/user-attachments/files/18222937/Test.App.Super.Table-export-1734893185470.enc.tar.gz)

**Reply by poirazis** [reply]
@dungtrantt the exported app is encoded, i am gonna need the pass or an unencoded one

**Reply by dungtrantt** [reply]
@poirazis Please use this, it wasn't encrypted
[Test App Super Table-export-1734962698492.tar.gz](https://github.com/user-attachments/files/18230138/Test.App.Super.Table-export-1734962698492.tar.gz)

**Reply by poirazis** [reply]
@dungtrantt Tested on both cloud and local dev, it seems to work fine here. Even added the binding in the request confirmation to make sure.

What budibase version are you using ? This was tested on latest released.

https://github.com/user-attachments/assets/db2871be-fcd4-4b85-b81d-aea390964a15

[Michaels app-export-1734973486841.tar.gz](https://github.com/user-attachments/files/18231706/Michaels.app-export-1734973486841.tar.gz)

**Reply by dungtrantt** [reply]
@poirazis The expectation is that data field in the result should have values of a dict, but you can see above video, it was empty.

**Reply by poirazis** [reply]
oh i think i found it @dungtrantt , the binding on the REST api query wasnt set up correctly.
it should have had doube braces to capture the value of the binding, not single

Change it to double curly braces and it works .
The result was empty cause when you run the query manually it was also empty.

![Screenshot 2024-12-23 at 7 52 57 PM](https://github.com/user-attachments/assets/c0b14369-efdd-44f2-a708-647916d57f70)

**Reply by dungtrantt** [reply]
@poirazis Awesome! It works now. Thank you so much. 
But, one more point, as I mentioned at the begining, we have Edit button in rows also. We expect this button would open an Modal with the data from selectedRow/ CurrentRow. How can we bind it? I did but it shows during binding/designing but it didn't work in Preview.

**Reply by poirazis** [reply]
@dungtrantt you shouldnt directly bind to the current row, as its dynamically enriched before action executions.
its meant to be a source to get values into state.

the selected row, is a static binding, i.e it keeps its value as long as the row is selected.

hope this helps

**Reply by dungtrantt** [reply]
@poirazis Thank you so much for your support. Everything works now.
